lib/fluent/plugin/buffer.rb

```diff
@@ -532,13 +491,16 @@ def purge_chunk(chunk_id)
 
           @dequeued_num[chunk.metadata] -= 1
           if metadata && !@stage[metadata] && (!@queued_num[metadata] || @queued_num[metadata] < 1) && @dequeued_num[metadata].zero?
-            @metadata_list.delete(metadata)
             @queued_num.delete(metadata)
             @dequeued_num.delete(metadata)
-            del_timekey(metadata)
           end
           log.trace "chunk purged", instance: self.object_id, chunk_id: dump_unique_id_hex(chunk_id), metadata: metadata
         end
+
+        if metadata && metadata.timekey
+          @mutex.synchronize { del_timekey(metadata.timekey) }
```

Comment:
`@mutex.synchronize` is called in `del_timekey`

Reply:
I forgot about it! thanks.
 .https://github.com/fluent/fluentd/pull/2563/commits/5c81b40df6586e0ad659ef045ea0536b969e13b2